Remove <<reset_all>> magic value from Settings

Fryguy opened the pull request:
~~Cleans up the changes in #17482~~

~~https://bugzilla.redhat.com/show_bug.cgi?id=1576984~~

~~WIP because the reset_all behavior doesn't make sense to me, and I'm trying to understand the use case for it.  I believe it needs to be changed completely.  Additionally there is a bug in that it will delete all settings across all regions in a multi-region database.~~

The user expectations around `<<reset_all>>` are too complicated, and so
this is being removed.  In doing so, the logic for `<<reset>>` has been
greatly simplified.

A different keyword is expected to implemented in a separate PR to deal
with enforcing a setting at the Region or Zone level and pushing that
value down to the child resources.  (This is not how `<<reset_all>>`
presently works, but that is a feature we expect to need)

https://bugzilla.redhat.com/show_bug.cgi?id=1576984



cc @yrudman @carbonin @gtanzillo

Fryguy commented on lib/vmdb/settings.rb:
```diff
@@ -189,6 +188,22 @@ def self.reset_settings_constant(settings)
+          owner[key] = parent_settings.fetch_path(*path)
+        when RESET_ALL_COMMAND
+          template_settings ||= build_template.load!.to_hash
+          owner[key] = template_settings.fetch_path(*path)
```
Ignore my implementation of reset_all, because I have no idea what the "real" reset_all is trying to do.

yrudman replied:
@Fryguy ```reset_all>>``` is trying to remove all changes to some key/node and bring ```Settings``` to the initial state (for that key/node) when values coming from ```yml``` file.

```<<reset_all>>``  is quite destructive command - remove all changes without back-up and may be we should just remove it.

For global region: ```setting_changes``` not excluded from replication and, I think, we should ignore both ```<<reset>>``` and ```<<reset_all>>``` on global region

carbonin replied:
@Fryguy I think the comment [here](https://github.com/ManageIQ/manageiq/pull/17563/files#diff-6bbc56fba708243deeb6a7af2b80b06aR26) does a fairly good job of describing what the original intention was.

I guess a better description would be to reset that key to the OOTB value. I think the confusion here is how to make that happen. If I understand correctly there are two options:
1. Explicitly set the value for this resource to the value from the `settings.yml`
    - This option doesn't affect the setting for any other resource under the same parent.
    - What this PR is doing (I think)
2. Remove all customizations for this key
    - This will change the setting for other resources (I don't think this would be do-able with the proposed implementation in this PR)
    - What the original PR did

I think the first option will be the least surprising to the user.

Fryguy replied:
Going to continue this conversation below, so as not to split the 2 threads (didn't see the responses here initially) - and also because this is going to get lost on code change

Fryguy commented on lib/vmdb/settings.rb:
```diff
@@ -189,6 +188,22 @@ def self.reset_settings_constant(settings)
+        case value
+        when RESET_COMMAND
+          parent_settings ||= parent_settings_without_local(resource).load!.to_hash
+          owner[key] = parent_settings.fetch_path(*path)
```
This passes all existing tests for the reset command, though I think a few more tests are needed.

Fryguy commented:
I'm also not sure about the names `<<reset>>` and `<<reset_all>>` in general.  I wonder if `<<inherit>>` is better, and I don't know what a replacement for `<<reset_all>>` would be yet.

yrudman commented:
should we use more words to be  more descriptive, something like
```take_from_parent``` for reset
and
```remove_all_changes_everywhere``` for reset_all (if we would decide to keep it)

carbonin commented:
> I wonder if <<inherit>> is better, and I don't know what a replacement for <<reset_all>> would be yet.

Maybe `<<inherit>>` and `<<default>>`? 
The only issue I see is `default` could be taken to mean either "get from parent" or "OOTB setting" unless you know that the other option is `inherit` which is quite clear.

Fryguy commented:
> remove_all_changes_everywhere for reset_all (if we would decide to keep it)

I still don't understand what this means.  Are you implying that if I'm on Server 1's settings screens, and I set it to <<reset_all>> then every server, region and zone in the entire database should be set to that value?  Or just server 1's zone and region?  Something else?

Fryguy commented:
I think `<<inherit>>` is super-clear compared to `<<reset>>`...the user just has to be aware of the hierarchy, which hopefully they are because of the tree-view on the left.

Fryguy commented:
> Continuing from ☝️ https://github.com/ManageIQ/manageiq/pull/17563#discussion_r194452342

As you stated, there are multiple ways to interpret what it means, which is why I don't like it given that it's so destructive.  As a user, if I'm on Server's 1's settings, I would never expect it to modify Server 5's settings (not to mention other zones or regions).  However, if I'm on Zone 1's settings, I could possibly see a feature where I want to "push down" the settings to all of the child servers (for that zone only), but I can't see how in a YAML interface you would set a value that you want pushed down and also put a magic string to say "push this down".  In that case, I would have a small but separate UI panel for forced inheritance.

I'm leaning towards removing the <<reset_all>> functionality entirely until we can understand the UX aspects of it.  @terezanovotna care to weigh in here?  I can brief you offline to catch you up if needed.

carbonin commented:
> I'm also not sure that 1 is the right thing either. If the OOtB value is x, and a Zone has declared y, then <<default>> (or whatever <<reset_all>> is called) on a server would set it back to x just for this server. Is that the behavior we want?

That's how I was interpreting it if not what the original PR did. Are there other options?

The comment in the BZ (from @gtanzillo) was this:
> A value of [default] would instruct the model to delete the existing "setting_changes" row for that key which would result in making the default values (from settings.yml) be the current value.

I think there's a bit of a misunderstanding here as removing "the existing setting_changes row for that key" would not make the OOTB setting the current value, it would be more like the "inherit" option we're discussing. I don't see any mention of anything which sounds like any of the "reset_all" situations we've been talking about.

Maybe the scope should be limited to the selected resource for now and we should spec out something new for altering multiple resources at once. Maybe that needs a new UX which isn't editing YAML? (:trollface: /2)

Fryguy commented:
I talked with @gtanzillo and the reason he suggested `<<reset>>` as the name over `<<inherit>>` is for the case where a user created a new key, and then wants to delete it.  It feels weird from the User perspective to say "inherit" when you want to delete, and `<<reset>>` seems to handle both cases.  Alternately, we could support both `<<inherit>>` and `<<delete>>`, and just implement them the same.  Given that, maybe `<<reset>>` is sufficient

I thought of a new way to do handle the feature of pushing down settings.  At a zone or region level, if the user changes a value to `<<enforce>>`, then we remove the setting on all resources in the subtree.  If entered at the server level, it would be a noop (or perhaps an error).

yrudman commented:
> At a zone or region level, if the user changes a value to <<enforce>>, then we remove the setting on all resources in the subtree. 

👍 

It is more logical to remove from subtree than from everywhere. And task to clean-up ```setting_changes``` table to original state (no records) could be accomplished in 2 steps: first is ```reset``` on region level and second is  ```enforce```.

Fryguy commented:
I discussed this with @carbonin yesterday (sorry @yrudman...didn't realize you were back :smile:), and we decided to just remove the `<<reset_all>>` functionality in this PR as it's just not something we'd expect a customer to understand.  The `<<enforce>>` feature will be done in a separate PR, mostly because it's a completely separate feature, but also because the placement of the code will likely be very different that the current code (probably going to live in ConfigurationManagementMixin instead of Vmdb::Settings).

Fryguy commented:
@carbonin @yrudman I think this is ready to go.  I updated the OP and PR title with the new description.

carbonin commented on spec/lib/vmdb/settings_spec.rb:
```diff
@@ -293,54 +293,32 @@
           expect(miq_server.settings_changes.find_by(:key => "/api/token_ttl")).to be nil
-          expect(Vmdb::Settings.for_resource(miq_server).api.new_key).to eq "new value"
-          expect(Vmdb::Settings.for_resource(second_server).api.token_ttl).to eq server_value
+          expect(described_class.for_resource(miq_server).api.new_key).to eq "new value"
```
I know that these specs haven't really changed, but why are we even testing this condition here?

yrudman replied:
@carbonin shouldn't we test that only specific key was deleted and other keys were not touched ?

carbonin commented on spec/lib/vmdb/settings_spec.rb:
```diff
@@ -293,54 +293,32 @@
-          expect(Vmdb::Settings.for_resource(miq_server).api.new_key).to eq "new value"
-          expect(Vmdb::Settings.for_resource(second_server).api.token_ttl).to eq server_value
+          expect(described_class.for_resource(miq_server).api.new_key).to eq "new value"
+          expect(described_class.for_resource(second_server).api.token_ttl).to eq server_value
```
Same thing here, why are we interested in the second server at all in this spec? Shouldn't we be testing this?:

```ruby
expect(described_class.for_resource(miq_server).api.token_ttl).to eq zone_value
```

carbonin commented on spec/lib/vmdb/settings_spec.rb:
```diff
@@ -293,54 +293,32 @@
-          expect(Vmdb::Settings.for_resource(miq_server).api.new_key).to eq nil
-          expect(Vmdb::Settings.for_resource(second_server).api.token_ttl).to eq server_value
+          expect(described_class.for_resource(miq_server).api.new_key).to eq nil
+          expect(described_class.for_resource(second_server).api.token_ttl).to eq server_value
```
Same here, why do we care about the second server?

carbonin commented on spec/lib/vmdb/settings_spec.rb:
```diff
@@ -293,54 +293,32 @@
-        it "passes validation" do
-          described_class.save!(miq_server, :session => {:timeout => reset_all})
-          expect(SettingsChange.where("key LIKE ?", "/session%").count).to eq 0
+          expect(described_class.for_resource(miq_server).session.timeout).to eq zone_value
```
This spec actually looks exactly like what I would expect `"deletes key-value for specific key for the resource if specified on leaf level"` to look like.

Fryguy commented:
@carbonin I'll be honest, I didn't even look at the tests, and just made sure they passed.  I'll double check their sensibility.

Fryguy commented:
@carbonin I reworked the tests and added a couple of new ones.  In doing so, there was actually a bug in the delete of a new key that was unseen, so 👍 .  I think I want to add a few more specs, particularly around that Array case, as I don't understand what's expected.

Fryguy commented:
@carbonin This should be good to go now.

Fryguy commented:
@carbonin @yrudman The logic in here piggybacks on how the HashDiffer works

So, the full algorithm is now
- take the sparsh hash and merge it into the full hash
- (new code) find any `<<reset>>` values and replace with the parent's value (inherit case); if the parent doesn't have a value, mark the value as `HashDiffer::MissingKey` (delete case).  We have to use MissingKey here, as opposed to nil, because the HashDiffer is aware of nils and will otherwise treat that as an actual value instead of a removal of a value.
- validate.  This will work because in the inherit case we should inherit already valid values, and in the delete case we can only delete values that never existed in the first place, so validation will ignore those.
- diff with the parent hash.  In the inherit case, the differ will only find the changes compared to the parent, but since they were copied from the parent, the desired state is now "no changes from the parent".  Similarly, in the delete case, the same things happens, it will see a MissingKey previously, and the new value is also MissingKey, so the desired state is now "no changes from the parent".  Thus, in both cases, the delta array willl not have an entry for that key.
- apply_settings_changes.  This works the same.  It takes the deltas, but sees there was a previous change, and now we want "no change", so it will delete the settings_change row.
